### PR TITLE
Invenio 3

### PR DESCRIPTION
**Pull request by jalavik**
```
* INCOMPATIBLE Makes `invenio-matcher` compatible with
  Invenio 3 release series.

* NEW Python 3 support.

* Updates tests and package configuration.
```

### CODE REVIEW

**Comment by jalavik** [comment]
@jacquerie coverage decreased due to mocking the `search` function calling `invenio_search.current_search_client.search` call which I was not able to mock successfully. Anyway, I think we should be okay to ship :ship:

**Comment by kaplun** [comment]
I like from the commit log:

> INCOMPATIBLE Makes `invenio-matcher` compatible with Invenio 3 release series.

**Comment by jalavik** [comment]
@kaplun Heh, yeah. :dancers:

File: invenio_matcher/api.py [hunk 1]
```diff
@@ -25,25 +25,21 @@
...
-    The API is agnostic of implementation-required parameters, such as
-    the `index` and the `doc_type` for Elasticsearch. The idea is to pass
-    them as keyword arguments: the implementation then passes them down to
-    the actual implementation in the engine.
+    The API requires `index` and the `doc_type` for Invenio-Search.
 
     NOTE: Callers might find tedious that they have to write boilerplate like
     ```
-    match(r, index='records', doc_type='record')
+    match(r, )
```
Comment (jacquerie) [inline invenio_matcher/api.py]: Nit: `match(r, index, doc_type)`, no?

File: invenio_matcher/core.py [hunk 2]
```diff
@@ -43,19 +46,35 @@ def execute(query, record, **kwargs):
...
+def get_queries(index, doc_type, **kwargs):
+    """Return queries defined for the given index and doc_type."""
+    MATCHER_QUERIES = current_app.config.get('MATCHER_QUERIES')
+
+    if not MATCHER_QUERIES or index not in MATCHER_QUERIES or \
+            doc_type not in MATCHER_QUERIES.get(index, []):
+        raise NoQueryDefined('No query defined for index {index} and doc_type'
+                             ' {doc_type} in MATCHER_QUERIES.'.format(
+                                 index=index, doc_type=doc_type))
+    return MATCHER_QUERIES[index][doc_type]
```
Comment (jacquerie) [inline invenio_matcher/core.py]: Nit: better style might be a try/catch?

File: invenio_matcher/errors.py [hunk 3]
```diff
@@ -25,17 +25,17 @@
 """Matcher errors."""
 
 
-class NoEngineDefined(KeyError):
-    pass
+class MatcherError(Exception):
+    """Base error class for Invenio-Matcher."""
```
Comment (jacquerie) [inline invenio_matcher/errors.py:29]: Not including this was probably my biggest mistake in designing this API. :+1:

File: tests/test_core.py [hunk 4]
```diff
@@ -24,181 +24,221 @@
...
-        query = {'type': 'exact', 'match': 'titles.title'}
-        record = Record(self.data)
+        assert result == expected
 
-        execute(query, record)
 
-        exact.assert_called_with(match='titles.title', values=['foo bar'])
+def test_execute_fuzzy(app, simple_record, mocker):
+    """Dispatch the query of type fuzzy."""
+    from invenio_records import Record
```
Comment (jacquerie) [inline tests/test_core.py]: Nit: imports inside of tests where needed to work around the previous registry madness, now they could be dropped.

**Comment by jacquerie** [comment]
Just nits, no blockers. If you want, fix them, but to me everything looks fine. :shipit:

**Pull request merged**